contents/security/sqlInjection.md

```diff
@@ -0,0 +1,145 @@
+```
+
+The SQL command string built from this input would be as follows:
+
+```html
+SELECT Username FROM Users WHERE Username = ‘foo’
+AND Password = ‘bar’ OR ‘’=’’
+```
+
+This query will always return true, as ‘’=’’ is always true, and it is connected using an 
+OR comparator. As such, the query can be simplified to this:
+
+```html
+SELECT Username FROM Users
+```
+
+This will return all the rows from the Users table in the database, regardless of username 
+or password entered. If used in user authentication, the user will likely be able to gain
+access to the system.
+
+
+## Methods of SQL Injection
+
+There are two common methods for attackers to identify and exploit an SQL injection weakness
+in websites:
+
+* SQL Injection - The attacker makes use of errors generated from executed queries. The 
+errors from the database allows the attacker to fine-tune their attack so their next 
+attempt would be more successful.
+
+For example, if when attempting this attack to retrieve data from a separate table:
+
+```html
+http://example/article.asp?ID=2+union+all+select+name+from+sysobjects
+```
+
+The attacker recieves a message similar to this:
+
+```html
+Microsoft OLE DB Provider for ODBC Drivers error 
+'80040e14' 
+[Microsoft][ODBC SQL Server Driver][SQL Server]All 
+queries in an SQL statement containing a UNION 
+operator must have an equal number of expressions
+in their target lists. 
+```
+
+Then the attacker would know that the separate table exists, and would only need 
+to change the number of columns retrieved to make the query successful.
+
+* Blind SQL Injection - Even when there is no detailed error from the database 
+displayed, attackers can gather information though other means, such as by testing 
+different ture or false statements.
+
+Executing the following statement:
+
+```html
+http://example/article.asp?ID=2+and+1=1
+```
+
+Should have the same result as the query below, because 1=1 is always true.
+
+```html
+http://example/article.asp?ID=2
+```
+
+However, if this statement was used instead:
+
+```html
+http://example/article.asp?ID=2+and+1=0
+```
+
+Since 1=0 is always false, the page should always return an error. If both queries 
+lead to the same result, then the website is probably safe from SQL injection.
+
+Once a SQL injection vulnerability is discovered, the attacker can continue to 
+exploit the database as per normal SQL Injection.
+
+
+## Preventing SQL Injection
+
+There are a couple of ways to protect your website against SQL injection attacks. The two most
+common ways are:
+
+* Using Prepared Statements using Parameterized Queries: By defining all the SQL code first 
+and passing in the parameters afterwards, you can make the database distinguish the difference 
+between code and data. It would treat the values entered by the user as a parameter, and would
+not allow it to alter the query being executed. The way to achieve this varies by language, 
+but it is easy to implement and effective.
```

Comment:
I think this thing is best explained by example. Just pick a server-side language that support SQL prepared statement and show the code difference between normal SQL queries and SQL prepared statement.